src/mapping/parser/grammar.pest

```diff
@@ -0,0 +1,91 @@
+        statement ~
+    NEWLINE* ~ "}" )?
+}
+
+// Paths
+path_index = ${ "[" ~ ('0'..'9')+ ~ "]" }
+path_segment = ${ ('a'..'z' | 'A'..'Z' | "_" | "-" )+ ~ path_index? }
+quoted_path_segment = ${ "\"" ~ inner_quoted_string ~ "\"" ~ path_index? }
+
+target_path = @{ ("." ~ (path_segment | quoted_path_segment))+ }
+
+// Functions
+function = { deletion | only_fields }
+
+deletion = { "del(" ~ target_path ~ ("," ~ target_path)* ~ ")" }
+only_fields = { "only_fields(" ~ target_path ~ ("," ~ target_path)* ~ ")" }
+
+// Queries
+path_coalesce = !{ "(" ~ (path_segment | quoted_path_segment) ~ ("|" ~ (path_segment | quoted_path_segment))+ ~ ")" }
+
+dot_path = ${ ("." ~ (path_segment | quoted_path_segment | path_coalesce))+ }
+
+// Query functions
+//
+// NOTE: Not including delete here yet, but there's no reason we can't slot it
+// in at some point and get `.nested = del(.foo)`, same with `only_fields`.
+query_function = { to_string | to_int | to_float | to_bool | to_timestamp }
+
+to_string = { "string(" ~ query_arithmetic ~ ("," ~ (string | null) )? ~ ")"}
+to_int = { "int(" ~ query_arithmetic ~ ("," ~ (number | null) )? ~ ")"}
+to_float = { "float(" ~ query_arithmetic ~ ("," ~ (number | null) )? ~ ")"}
+to_bool = { "bool(" ~ query_arithmetic ~ ("," ~ (boolean | null) )? ~ ")"}
+to_timestamp = { "timestamp(" ~ query_arithmetic ~ "," ~ string ~ ")"}
+
+group = { "(" ~ query_arithmetic ~ ")" }
+
+value = _{ string | number | boolean | null }
+
+boolean = { "true" | "false" }
+
+null = { "null" }
+
+string = ${ "\"" ~ inner_quoted_string ~ "\"" }
+inner_quoted_string = @{ char* }
+char = {
+    !("\"" | "\\") ~ ANY
+    | "\\" ~ ("\"" | "\\" | "n" | "t")
+    // TODO: Eventually expand this to allow more escape sequences.
+    // | "\\" ~ ("\"" | "\\" | "/" | "b" | "f" | "n" | "r" | "t")
+    // | "\\" ~ ("u" ~ ASCII_HEX_DIGIT{4})
+}
+
+number = @{
+    "-"?
+    ~ ("0" | ASCII_NONZERO_DIGIT ~ ASCII_DIGIT*)
+    ~ ("." ~ ASCII_DIGIT*)?
+    ~ (^"e" ~ ("+" | "-")? ~ ASCII_DIGIT+)?
+}
+
+query_leaf = _{ value | dot_path | group | query_function }
+
+// Arithmetic, broken down into tiers in order to support operator precedence.
+// Operators of the same tier are resolved from left to right.
+
+// Deepest tier is the highest order of precedence, we call this just 'query' as
+// it'll appear in parser error messages.
+arithmetic_operator_product = { "*" | "/" | "%" }
+query = { query_leaf ~ (arithmetic_operator_product ~ query_leaf)* }
+
+arithmetic_operator_sum = { "+" | "-" }
+query_arithmetic_sum = { query ~ (arithmetic_operator_sum ~ query)* }
+
+arithmetic_operator_compare = { ">=" | ">" | "<=" | "<" | "!=" | "==" }
+query_arithmetic_compare = { query_arithmetic_sum ~ (arithmetic_operator_compare ~ query_arithmetic_sum)* }
+
+arithmetic_operator_boolean = { "||" | "&&" }
+query_arithmetic_boolean = { query_arithmetic_compare ~ (arithmetic_operator_boolean ~ query_arithmetic_compare)* }
+
+query_arithmetic = _{ query_arithmetic_boolean }
```

Comment:
I'm not seeing any way to express unary operations such as negation or (boolean or bitwise) inversion. That seems a significant omission if it isn't planned.